File: windows/ChoroReader.Core/UnigramTokenizer.cs
using System.Text;
using System.Text.Json;

namespace ChoroReader.Core;

/// <summary>
/// Unigram（SentencePiece）方式のトークナイザ。
///
/// <para>
/// Ruri v3 の <c>tokenizer.json</c> をそのまま読む。30m / 130m / 310m で同じファイルなので、
/// モデルの大きさを替えてもここは替わらない。
/// </para>
/// <para>
/// 段は 4 つ。
/// <list type="number">
/// <item>特殊トークン（added_tokens）で切り分ける。中身は素通し</item>
/// <item>Metaspace：空白を ▁ に置き換える（prepend_scheme=never なので頭には足さない）</item>
/// <item>Viterbi：語彙の並びで最も点の高い分け方を選ぶ</item>
/// <item>語彙に無い文字は byte fallback で <c>&lt;0xXX&gt;</c> に落とす</item>
/// </list>
/// </para>
/// <para>
/// <b>既製品では揃わない。</b><c>Microsoft.ML.Tokenizers</c> を凍結済みの期待値に通したところ
/// 31/37 で、残ったうち 1 件が Viterbi の同点の採り方の違いだった（外からは直せない）。
/// 測った結果は spikes/findings-csharp-tokenizer.md にある。
/// </para>
/// <para>
/// <b>食い違っても例外は出ず、黙って違うベクトルになる。</b>参照実装（kohagi、
/// Rust の tokenizers crate）とトークン ID が一致することを検査で固める以外に、
/// 守る手立てが無い。Swift 実装は次の 3 つで黙って間違えた
/// （spikes/findings-swift-tokenizer.md）。移すにあたって同じ 3 つを踏まないようにしてある。
/// </para>
/// <list type="bullet">
/// <item>
/// <b>語彙は文字列でなくバイト列で引く。</b>Swift の String は正準等価で比べるため
/// 「が」（1 符号位置）と「か」＋濁点（2 符号位置）が同じ鍵になり、片方が消えた。
/// C# の string は序数で比べるのでそこは踏まないが、JSON から来た並びを
/// UTF-16 へ通して戻す往復そのものを避けたいので、バイト列のまま扱う
/// </item>
/// <item>
/// <b>点の足し込みは double で行う。</b>語彙の点は f32 だが参照は f64 で積む。
/// 同じ点の分け方が並ぶと（目次の点線のような連なり）丸めで順位が入れ替わる
/// </item>
/// <item>
/// <b>同点は開始位置の小さい方（長い語彙）を採る。</b>バイト位置を昇順に進め、
/// 等号を含まない比較で更新すれば参照と同じになる
/// </item>
/// </list>
/// </summary>
public sealed class UnigramTokenizer
{
    private readonly record struct Entry(int Id, double Score);

    /// <summary>語彙。UTF-8 のバイト列から引く。</summary>
    private readonly Dictionary<byte[], Entry> _vocab;

    /// <summary>
    /// 走査の途中でバイト列を切り出さずに引くための口。
    ///
    /// <para>
    /// Viterbi は 1 文字進むごとに語彙を何度も引く。そのたびに配列を作ると、
    /// 1 冊（580 段落）で数百万回の割り当てになる。
    /// </para>
    /// </summary>
    private readonly Dictionary<byte[], Entry>.AlternateLookup<ReadOnlySpan<byte>> _lookup;

    private readonly int _unkId;
    private readonly bool _byteFallback;

    /// <summary>0…255 → <c>&lt;0xXX&gt;</c> の id。無ければ -1。</summary>
    private readonly int[] _byteToken;

    /// <summary>特殊トークン。長い順に見る。</summary>
    private readonly (byte[] Content, int Id)[] _added;

    private readonly int _bos;
    private readonly int _eos;
    private readonly double _minScore;

    /// <summary>語彙に載る最長のバイト数。Viterbi の探索幅を切る。</summary>
    private readonly int _maxPieceBytes;

    /// <summary>語彙に無い 1 文字に与える点。tokenizers の K_UNK_PENALTY と同じ。</summary>
    private const double UnkPenalty = 10.0;

    /// <summary>Metaspace の印（U+2581）。</summary>
    private static ReadOnlySpan<byte> Mark => "▁"u8;

    public static UnigramTokenizer Load(string path) =>
        new(JsonDocument.Parse(File.ReadAllBytes(path)));

    public static UnigramTokenizer Parse(string json) =>
        new(JsonDocument.Parse(json));

    private UnigramTokenizer(JsonDocument document)
    {
        using var owned = document;
        var root = owned.RootElement;

        if (!root.TryGetProperty("model", out var model)
            || !model.TryGetProperty("vocab", out var vocab)
            || vocab.ValueKind != JsonValueKind.Array
            || vocab.GetArrayLength() == 0)
        {
            throw new DocumentException("cannotReadTokenizer", "model.vocab がありません");
        }

        _vocab = new Dictionary<byte[], Entry>(vocab.GetArrayLength(), new ByBytes());
        var longest = 0;
        var lowest = double.MaxValue;
        var id = 0;
        foreach (var entry in vocab.EnumerateArray())
        {
            if (entry.ValueKind != JsonValueKind.Array || entry.GetArrayLength() < 2)
            {
                throw new DocumentException("cannotReadTokenizer", $"語彙 {id} の形が違います");
            }
            var piece = entry[0].GetString()
                        ?? throw new DocumentException("cannotReadTokenizer", $"語彙 {id} の形が違います");
            var score = entry[1].GetDouble();

            var bytes = Encoding.UTF8.GetBytes(piece);
            _vocab[bytes] = new Entry(id, score);
            longest = Math.Max(longest, bytes.Length);
            lowest = Math.Min(lowest, score);
            id++;
        }
        _lookup = _vocab.GetAlternateLookup<ReadOnlySpan<byte>>();
        _maxPieceBytes = longest;
        _minScore = lowest;

        _unkId = model.TryGetProperty("unk_id", out var unk) && unk.ValueKind == JsonValueKind.Number
            ? unk.GetInt32()
            : 0;
        _byteFallback = model.TryGetProperty("byte_fallback", out var fallback)
                        && fallback.ValueKind == JsonValueKind.True;

        _byteToken = new int[256];
        for (var b = 0; b < 256; b++)
        {
            _byteToken[b] = _lookup.TryGetValue(Encoding.UTF8.GetBytes($"<0x{b:X2}>"), out var found)
                ? found.Id
                : -1;
        }

        // 短いものが先に当たると取りこぼす。長い順に見る。
        _added = root.TryGetProperty("added_tokens", out var addedTokens)
                 && addedTokens.ValueKind == JsonValueKind.Array
            ? [.. addedTokens.EnumerateArray()
                .Where(t => t.TryGetProperty("content", out _) && t.TryGetProperty("id", out _))
                .Select(t => (Content: Encoding.UTF8.GetBytes(t.GetProperty("content").GetString() ?? string.Empty),
                              Id: t.GetProperty("id").GetInt32()))
                .OrderByDescending(t => t.Content.Length)]
            : [];

        // post_processor は TemplateProcessing で <s> … </s> を足すだけ。
        _bos = SpecialId(root, "<s>");
        _eos = SpecialId(root, "</s>");
    }

    private static int SpecialId(JsonElement root, string name)
    {
        if (root.TryGetProperty("post_processor", out var post)
            && post.TryGetProperty("special_tokens", out var special)
            && special.TryGetProperty(name, out var one)
            && one.TryGetProperty("ids", out var ids)
            && ids.ValueKind == JsonValueKind.Array
            && ids.GetArrayLength() > 0)
        {
            return ids[0].GetInt32();
        }
        return -1;
    }

    /// <summary>
    /// 文字列をトークン ID の並びにする。
    /// <paramref name="addSpecialTokens"/> が真なら <c>&lt;s&gt;</c> … <c>&lt;/s&gt;</c> で挟む。
    /// </summary>
    public IReadOnlyList<int> Encode(string text, bool addSpecialTokens = true)
    {
        var out_ = new List<int>();
        if (addSpecialTokens && _bos >= 0)
        {
            out_.Add(_bos);
        }

        var bytes = Encoding.UTF8.GetBytes(text);
        foreach (var piece in SplitOnAdded(bytes))
        {
            if (piece.Id >= 0)
            {
                out_.Add(piece.Id);
            }
            else
            {
                Viterbi(Metaspace(bytes.AsSpan(piece.From, piece.To - piece.From)), out_);
            }
        }

        if (addSpecialTokens && _eos >= 0)
        {
            out_.Add(_eos);
        }
        return out_;
    }

    // MARK: 段 1：特殊トークンで切り分ける

    /// <summary>切り分けた一片。<see cref="Id"/> が 0 以上なら特殊トークン、負なら素の範囲。</summary>
    private readonly record struct Split(int From, int To, int Id);

    private List<Split> SplitOnAdded(byte[] bytes)
    {
        var out_ = new List<Split>();
        if (_added.Length == 0)
        {
            if (bytes.Length > 0)
            {
                out_.Add(new Split(0, bytes.Length, -1));
            }
            return out_;
        }

        var at = 0;
        var plainFrom = 0;
        while (at < bytes.Length)
        {
            var hit = -1;
            var length = 0;
            foreach (var (content, id) in _added)
            {
                if (at + content.Length <= bytes.Length
                    && bytes.AsSpan(at, content.Length).SequenceEqual(content))
                {
                    hit = id;
                    length = content.Length;
                    break;
                }
            }
            if (hit >= 0)
            {
                if (at > plainFrom)
                {
                    out_.Add(new Split(plainFrom, at, -1));
                }
                out_.Add(new Split(at, at + length, hit));
                at += length;
                plainFrom = at;
            }
            else
            {
                at++;
            }
        }
        if (plainFrom < bytes.Length)
        {
            out_.Add(new Split(plainFrom, bytes.Length, -1));
        }
        return out_;
    }

    // MARK: 段 2：Metaspace

    /// <summary>空白を ▁ に置き換える。prepend_scheme が never なので頭には足さない。</summary>
    private static byte[] Metaspace(ReadOnlySpan<byte> bytes)
    {
        var out_ = new List<byte>(bytes.Length);
        foreach (var b in bytes)
        {
            if (b == 0x20)
            {
                out_.AddRange(Mark);
            }
            else
            {
                out_.Add(b);
            }
        }
        return [.. out_];
    }

    // MARK: 段 3：Viterbi

    /// <summary>最も点の高い分け方を選ぶ。バイト位置で進み、文字の頭だけを節点にする。</summary>
    private void Viterbi(byte[] bytes, List<int> into)
    {
        var count = bytes.Length;
        if (count == 0)
        {
            return;
        }

        // 継続バイト（10xxxxxx）は節点にしない。
        var boundary = new bool[count + 1];
        boundary[count] = true;
        for (var i = 0; i < count; i++)
        {
            boundary[i] = (bytes[i] & 0xC0) != 0x80;
        }

        var best = new double[count + 1];
        var from = new int[count + 1];
        // その節点へ入った語彙 id。-1 は語彙に無かった 1 文字。
        var arrived = new int[count + 1];
        Array.Fill(best, -double.MaxValue);
        Array.Fill(from, -1);
        Array.Fill(arrived, -1);
        best[0] = 0;

        var unkScore = _minScore - UnkPenalty;

        for (var i = 0; i < count; i++)
        {
            if (!boundary[i] || best[i] <= -double.MaxValue)
            {
                continue;
            }

            var matchedWholeChar = false;
            var limit = Math.Min(count, i + _maxPieceBytes);
            var width = CharLength(bytes[i]);

            for (var j = i + 1; j <= limit; j++)
            {
                if (!boundary[j] || !_lookup.TryGetValue(bytes.AsSpan(i, j - i), out var found))
                {
                    continue;
                }
                var value = best[i] + found.Score;
                // **等号を含めない。** 同点なら開始位置の小さい方（長い語彙）が残る。
                if (value > best[j])
                {
                    best[j] = value;
                    from[j] = i;
                    arrived[j] = found.Id;
                }
                if (j - i == width)
                {
                    matchedWholeChar = true;
                }
            }

            // その位置の 1 文字が語彙に無ければ、罰つきで進める（あとで byte fallback にする）
            if (!matchedWholeChar && i + width <= count)
            {
                var j = i + width;
                var value = best[i] + unkScore;
                if (value > best[j])
                {
                    best[j] = value;
                    from[j] = i;
                    arrived[j] = -1;
                }
            }
        }

        // 後ろから辿る。組み立ては逆順になるので、最後にひっくり返す。
        var reversed = new List<int>();
        var at2 = count;
        while (at2 > 0)
        {
            var previous = from[at2];
            if (previous < 0)
            {
                break;  // 進めなかった。起きないはずだが、途中まで返す
            }
            if (arrived[at2] >= 0)
            {
                reversed.Add(arrived[at2]);
            }
            else if (_byteFallback)
            {
                // 語彙に無い 1 文字は、UTF-8 のバイトごとに落とす
                for (var b = at2 - 1; b >= previous; b--)
                {
                    var token = _byteToken[bytes[b]];
                    reversed.Add(token >= 0 ? token : _unkId);
                }
            }
            else
            {
                reversed.Add(_unkId);
            }
            at2 = previous;
        }

        for (var i = reversed.Count - 1; i >= 0; i--)
        {
            into.Add(reversed[i]);
        }
    }

    /// <summary>そのバイトから始まる 1 文字のバイト数。</summary>
    private static int CharLength(byte first) => first switch
    {
        < 0x80 => 1,
        _ when (first & 0xE0) == 0xC0 => 2,
        _ when (first & 0xF0) == 0xE0 => 3,
        _ when (first & 0xF8) == 0xF0 => 4,
        _ => 1,
    };

    /// <summary>
    /// バイト列を鍵にするための比べ方。
    ///
    /// <para>
    /// 切り出さずに引けるよう、<c>ReadOnlySpan&lt;byte&gt;</c> からも引ける口を持たせる。
    /// </para>
    /// </summary>
    private sealed class ByBytes
        : IEqualityComparer<byte[]>, IAlternateEqualityComparer<ReadOnlySpan<byte>, byte[]>
    {
        public bool Equals(byte[]? left, byte[]? right) =>
            left is null ? right is null : right is not null && left.AsSpan().SequenceEqual(right);

        public int GetHashCode(byte[] key) => GetHashCode(key.AsSpan());

        public bool Equals(ReadOnlySpan<byte> span, byte[] key) => span.SequenceEqual(key);

        public int GetHashCode(ReadOnlySpan<byte> span)
        {
            var hash = new HashCode();
            hash.AddBytes(span);
            return hash.ToHashCode();
        }

        public byte[] Create(ReadOnlySpan<byte> span) => span.ToArray();
    }
}

File: macos/Sources/ChoroReader/Embedding/UnigramTokenizer.swift
import Foundation

struct UnigramTokenizer {
    /// 語彙。UTF-8 のバイト列から引く。
    private let scores: [[UInt8]: (id: Int, score: Double)]
    private let unkId: Int
    private let byteFallback: Bool
    private let byteToken: [Int]
    private let added: [(content: [UInt8], id: Int)]
    private let bos: Int
    private let eos: Int
    private let minScore: Double
    /// 語彙に載る最長のバイト数。Viterbi の探索幅を切る。
    private let maxPieceBytes: Int

    private static let unkPenalty: Double = 10.0

    enum Failure: Error, LocalizedError {
        case cannotRead(String)

        var errorDescription: String? {
            switch self {
            case let .cannotRead(why): return "トークナイザを読めません：\(why)"
            }
        }
    }

    init(contentsOf url: URL) throws {
        guard let json = try JSONSerialization.jsonObject(with: Data(contentsOf: url)) as? [String: Any],
              let model = json["model"] as? [String: Any],
              let vocab = model["vocab"] as? [[Any]], !vocab.isEmpty
        else { throw Failure.cannotRead("model.vocab がありません") }

        var table: [[UInt8]: (Int, Double)] = [:]
        table.reserveCapacity(vocab.count)
        var longest = 0
        var lowest = Double.greatestFiniteMagnitude
        for (id, entry) in vocab.enumerated() {
            guard entry.count >= 2, let piece = entry[0] as? String,
                  let score = (entry[1] as? NSNumber)?.doubleValue
            else { throw Failure.cannotRead("語彙 \(id) の形が違います") }
            let bytes = Array(piece.utf8)
            table[bytes] = (id, score)
            longest = max(longest, bytes.count)
            lowest = min(lowest, score)
        }
        scores = table
        maxPieceBytes = longest
        minScore = lowest
        unkId = (model["unk_id"] as? Int) ?? 0
        byteFallback = (model["byte_fallback"] as? Bool) ?? false
        byteToken = (0 ... 255).map { table[Array(String(format: "<0x%02X>", $0).utf8)]?.0 ?? -1 }

        added = ((json["added_tokens"] as? [[String: Any]]) ?? [])
            .compactMap { entry in
                guard let content = entry["content"] as? String, let id = entry["id"] as? Int
                else { return nil }
                return (Array(content.utf8), id)
            }
            .sorted { $0.0.count > $1.0.count }

        let post = (json["post_processor"] as? [String: Any]) ?? [:]
        let special = (post["special_tokens"] as? [String: Any]) ?? [:]
        func specialId(_ name: String) -> Int {
            guard let entry = special[name] as? [String: Any],
                  let ids = entry["ids"] as? [Int], let first = ids.first else { return -1 }
            return first
        }
        bos = specialId("<s>")
        eos = specialId("</s>")
    }

    func encode(_ text: String, addSpecialTokens special: Bool = true) -> [Int] {
        var out: [Int] = []
        if special, bos >= 0 { out.append(bos) }
        for piece in splitOnAdded(Array(text.utf8)) {
            switch piece {
            case let .plain(bytes): out.append(contentsOf: viterbi(metaspace(bytes)))
            case let .special(id): out.append(id)
            }
        }
        if special, eos >= 0 { out.append(eos) }
        return out
    }

    func encodePair(_ first: String, _ second: String, limit: Int) -> [Int] {
        var a = encode(first, addSpecialTokens: false)
        var b = encode(second, addSpecialTokens: false)
        var over = a.count + b.count + 4 - limit
        while over > 0, !(a.isEmpty && b.isEmpty) {
            if a.count > b.count { a.removeLast() } else { b.removeLast() }
            over -= 1
        }
        var out: [Int] = []
        out.reserveCapacity(a.count + b.count + 4)
        if bos >= 0 { out.append(bos) }
        out.append(contentsOf: a)
        if eos >= 0 { out.append(eos) }
        if bos >= 0 { out.append(bos) }
        out.append(contentsOf: b)
        if eos >= 0 { out.append(eos) }
        return out
    }

    // MARK: - 段 1：特殊トークンで切り分ける

    private enum Piece {
        case plain([UInt8])
        case special(Int)
    }

    private func splitOnAdded(_ bytes: [UInt8]) -> [Piece] {
        guard !added.isEmpty else { return bytes.isEmpty ? [] : [.plain(bytes)] }
        var out: [Piece] = []
        var at = 0
        var plainFrom = 0
        while at < bytes.count {
            var hit: (length: Int, id: Int)?
            for (content, id) in added where at + content.count <= bytes.count {
                if Array(bytes[at ..< at + content.count]) == content {
                    hit = (content.count, id)
                    break
                }
            }
            if let hit {
                if at > plainFrom { out.append(.plain(Array(bytes[plainFrom ..< at]))) }
                out.append(.special(hit.id))
                at += hit.length
                plainFrom = at
            } else {
                at += 1
            }
        }
        if plainFrom < bytes.count { out.append(.plain(Array(bytes[plainFrom...]))) }
        return out
    }

    // MARK: - 段 2：Metaspace

    /// 空白を ▁ に置き換える。prepend_scheme が never なので頭には足さない。
    private func metaspace(_ bytes: [UInt8]) -> [UInt8] {
        var out: [UInt8] = []
        out.reserveCapacity(bytes.count)
        let mark = Array("\u{2581}".utf8)   // E2 96 81
        for byte in bytes {
            if byte == 0x20 { out.append(contentsOf: mark) } else { out.append(byte) }
        }
        return out
    }

    // MARK: - 段 3：Viterbi

    private func viterbi(_ bytes: [UInt8]) -> [Int] {
        guard !bytes.isEmpty else { return [] }
        let count = bytes.count

        var isBoundary = [Bool](repeating: false, count: count + 1)
        isBoundary[count] = true
        for i in 0 ..< count where bytes[i] & 0xC0 != 0x80 { isBoundary[i] = true }

        var best = [Double](repeating: -.greatestFiniteMagnitude, count: count + 1)
        var from = [Int](repeating: -1, count: count + 1)
        var arrived = [Int](repeating: -1, count: count + 1)
        best[0] = 0

        let unkScore = minScore - Self.unkPenalty

        for i in 0 ..< count where isBoundary[i] && best[i] > -.greatestFiniteMagnitude {
            var matchedWholeChar = false
            let limit = min(count, i + maxPieceBytes)
            let width = charLength(bytes, i)
            var j = i + 1
            while j <= limit {
                if isBoundary[j], let (id, score) = scores[Array(bytes[i ..< j])] {
                    let value = best[i] + score
                    if value > best[j] { best[j] = value; from[j] = i; arrived[j] = id }
                    if j - i == width { matchedWholeChar = true }
                }
                j += 1
            }
            if !matchedWholeChar, i + width <= count {
                let j = i + width
                let value = best[i] + unkScore
                if value > best[j] { best[j] = value; from[j] = i; arrived[j] = -1 }
            }
        }

        var ids: [Int] = []
        var at = count
        while at > 0 {
            let previous = from[at]
            if previous < 0 { break }   // 進めなかった。起きないはずだが、途中まで返す
            if arrived[at] >= 0 {
                ids.append(arrived[at])
            } else if byteFallback {
                for byte in bytes[previous ..< at].reversed() {
                    let id = byteToken[Int(byte)]
                    ids.append(id >= 0 ? id : unkId)
                }
            } else {
                ids.append(unkId)
            }
            at = previous
        }
        return ids.reversed()
    }

    /// そのバイト位置から始まる 1 文字のバイト数。
    private func charLength(_ bytes: [UInt8], _ i: Int) -> Int {
        let byte = bytes[i]
        if byte < 0x80 { return 1 }
        if byte & 0xE0 == 0xC0 { return 2 }
        if byte & 0xF0 == 0xE0 { return 3 }
        if byte & 0xF8 == 0xF0 { return 4 }
        return 1
    }
}
